[1.8] Automated cherry pick of #58720 #57326 #60342

joelsmith opened the pull request:
Cherry pick of #58720 #57326 #60342 on release-1.8.

#58720: Ensure that the runtime mounts RO volumes read-only
#57326: Update Dashboard version to v1.8.3
#60342: Fix nested volume mounts for read-only API data volumes

Fixes https://github.com/kubernetes/kubernetes/issues/60814 for 1.8

Note for reviewers:
The Dashboard update is necessary because previous versions of the Dashboard attempted to write to the secret volume on startup, and so the read-only secret volume change prevented the Dashboard from starting. 

```release-note
Changes secret, configMap, downwardAPI and projected volumes to mount read-only, instead of allowing applications to write data and then reverting it automatically. Until version 1.11, setting the feature gate ReadOnlyAPIDataVolumes=false will preserve the old behavior. Updates dashboard version to v1.8.3 to avoid writing to read-only locations.
```

joelsmith committed:
Ensure that the runtime mounts RO volumes read-only

Add a feature gate ReadOnlyAPIDataVolumes to a provide a way to
disable the new behavior in 1.10, but for 1.11, the new
behavior will become non-optional.

Also, update E2E tests for downwardAPI and projected volumes
to mount the volumes somewhere other than /etc.

Commit:
Update Dashboard version to v1.8.3

joelsmith committed:
Fix nested volume mounts for read-only API data volumes

Since the runtime may try to create mount points within
the sandbox, it will fail if the mount point is within
a read-only API data volume, like a secret or configMap
volume.

Create any needed mount points during volume setup.

rata commented on cluster/addons/dashboard/dashboard-controller.yaml:
```diff
@@ -56,8 +56,7 @@ spec:
       - name: kubernetes-dashboard-certs
-        secret:
-          secretName: kubernetes-dashboard-certs
+        emptyDir: {}
```
I see there was an issue opened on the dashboard discussed before, https://github.com/kubernetes/kubernetes/pull/58720/files#r166033248, but it's better to back port like this, right?

rata replied:
Oh, now I see the other commit using the new dashboard. Dismiss this, please :)

rata commented on pkg/volume/configmap/configmap.go:
```diff
@@ -194,6 +194,9 @@ func (b *configMapVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
 	if err := wrapped.SetUpAt(dir, fsGroup); err != nil {
 		return err
 	}
+	if err := volumeutil.MakeNestedMountpoints(b.volName, dir, b.pod); err != nil {
```
Seems disturbing to use `b` here and `*b` here: https://github.com/kubernetes/kubernetes/pull/60515/commits/25f34665f8127ec3466c06e733da423772a17f48#diff-a3b28c0b2e7de9ad275e0880d0764c8dR187

But it's like the original PR, so perfect :)

rata reviewed (approved):
LGTM
Thanks a lot!

I reviewed commit by commit, for more easy inspect what belongs to which PR, and some lines were later modified in another commit. Sorry for the comments that were stale because the new commit fixed it :)

rata commented:
/lgtm

rata commented:
/assign @smarterclayton

joelsmith commented:
/kind bug
/sig storage
/sig node
/priority important-soon

msau42 commented:
/lgtm

joelsmith commented:
/retest

feiskyer reviewed (commented):
hack/verify-bazel.sh failed, need to run hack/update-bazel.sh

joelsmith commented:
/retest
Now that #60523 is merged, maybe it'll pass

liggitt commented:
/retest

joelsmith commented:
I didn't pay enough attention on the backport of the dashboard update. One of its changes prevented the dashboard from starting on 1.8:
```patch
-apiVersion: apps/v1beta2
+apiVersion: apps/v1
```
So I've reverted that change (and squashed the revert into the commit)

Also, the failing test seems to have passed all the test cases but had some problem during the teardown of the test cluster.
/retest

liggitt commented:
/status approved-for-milestone

smarterclayton commented:
/approve

BenTheElder commented:
Apologies, we had a small migration that required restarting the queue (`pull-kubernetes-unit` -> `pull-kubernetes-integration`), we've been waiting for a quiet enough day. See: https://github.com/kubernetes/test-infra/pull/7057#issuecomment-369384392

Getting this back on track for merge:
/test pull-kubernetes-integration

BenTheElder commented:
/retest

BenTheElder commented:
/test pull-kuberentes-node-e2e

BenTheElder commented:
/test pull-kubernetes-node-e2e

BenTheElder commented:
PR is now back in the queue (and next)
